### PR TITLE
resolvehelper(): remove goto's

### PR DESCRIPTION
**Pull request by WalterBright**
Nested functions are handy for removing goto's

### CODE REVIEW

File: src/dmd/typesem.d [hunk 1]
```diff
@@ -237,9 +244,26 @@ private void resolveHelper(TypeQualified mt, const ref Loc loc, Scope* sc, Dsymb
                 *pt = Type.terror;
                 return;
             }
+
+            void helper3()
```
Comment (JinShil) [inline src/dmd/typesem.d]: I suggest using a more descriptive name to help self-document the code.  I would suggest a name, but I'm afraid I can't decipher what it's doing, further motivating the need for a descriptive name.

File: src/dmd/typesem.d [hunk 2]
```diff
@@ -165,10 +165,17 @@ private void resolveTupleIndex(const ref Loc loc, Scope* sc, Dsymbol s, Expressi
...
- * Output:
- *      if expression, *pe is set
- *      if type, *pt is set
+ * it represents a Type, Expression, or Dsymbol.
+ * Params:
+ *      mt = array of identifiers
+ *      loc = location for error messages
+ *      sc = context
+ *      s = symbol to start search at
+ *      scopesym = unused
```
Comment (jacob-carlborg) [inline src/dmd/typesem.d]: If this is not used, perhaps remove the parameter name.

File: src/dmd/typesem.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/typesem.d]: I don't really know what it's doing, either.

Comment (WalterBright) [reply src/dmd/typesem.d]: I just try to leave things better than I found them.

File: src/dmd/typesem.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/typesem.d]: Perhaps, but I made enough changes for this PR.

File: src/dmd/typesem.d (hunk 1 above)
Comment (JinShil) [reply src/dmd/typesem.d]: > I don't really know what it's doing, either.

At first I laughed, but now I realize that's not funny.

Comment (WalterBright) [reply src/dmd/typesem.d]: `resolveHelper` is also a most unhelpful name. But I wrote the darn thing 20 years ago, and it'll take more work than this PR to make it a virtuous piece of code.

Comment (dejlek) [reply src/dmd/typesem.d]: :)

File: src/dmd/typesem.d [hunk 3]
```diff
@@ -237,9 +244,26 @@ private void resolveHelper(TypeQualified mt, const ref Loc loc, Scope* sc, Dsymb
                 *pt = Type.terror;
                 return;
             }
+
+            void helper3()
+            {
+                Expression e;
+                VarDeclaration v = s.isVarDeclaration();
```
Comment (dejlek) [inline src/dmd/typesem.d]: I am used to isSomeType() returning a boolean, which is not the case here... But I can't suggest a better name than getVarDeclaration() so I guess isVarDeclaration() is as good as ahything else... Same goes for all similar methods.

File: src/dmd/typesem.d (hunk 1 above)
Comment (JinShil) [reply src/dmd/typesem.d]: `symbolToExpression` maybe?

File: src/dmd/typesem.d (hunk 3 above)
Comment (JinShil) [reply src/dmd/typesem.d]: Yes, the `isSomeType` methods in DMD aren't the typical `isSomeType`, but we'd have to make a multi-thousand line PR to DMD to remedy that, and @WalterBright  has forbade that.  (See https://github.com/dlang/dmd/pull/7747).  That PR actually prompted @andralex to give me a call on my personal phone.  Yes, the great Andrei Alexandrescu called me... to scold me, that is.  One of the low moments in my D career, I must say.

File: src/dmd/typesem.d [hunk 4]
```diff
@@ -248,7 +272,7 @@ private void resolveHelper(TypeQualified mt, const ref Loc loc, Scope* sc, Dsymb
                     // https://issues.dlang.org/show_bug.cgi?id=13087
                     // this.field is not constant always
                     if (!v.isThisDeclaration())
-                        goto L3;
+                        return helper3();
```
Comment (JinShil) [inline src/dmd/typesem.d]: Why is this returning a `void` function?  Is this just another way of writing
```D
helper3();
return;
```

**Review by JinShil — APPROVED** [review]
I don't see any technical issues with this PR, and based on past experience @WalterBright seldom accommodates reviews based on craftsmanship and style.  The code is this PR is better than what's currently there, and we do have more important things to do.

**Review by jacob-carlborg — APPROVED** [review]

File: src/dmd/typesem.d (hunk 3 above)
Comment (andralex) [reply src/dmd/typesem.d]: @JinShil well as I remember it it was more for planning future work than that PR. Sorry I came across as negative!

File: src/dmd/typesem.d (hunk 4 above)
Comment (andralex) [reply src/dmd/typesem.d]: Yeah this is an idiom inherited from C++.

File: src/dmd/typesem.d (hunk 3 above)
Comment (JinShil) [reply src/dmd/typesem.d]: I know you were just trying to help me understand your expectations.  No offense was taken.

File: src/dmd/typesem.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/typesem.d]:
```
return helper3();
```
is more succint than:
```
{
    helper3();
    return;
}
```

File: src/dmd/typesem.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/typesem.d]: That's not at all what it does.

**Pull request merged**